### PR TITLE
pkg: truncindex: strongly type errors

### PR DESCRIPTION
**Pull request by runcom**
Signed-off-by: Antonio Murdaca amurdaca@redhat.com

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -141,13 +141,12 @@ func (daemon *Daemon) Get(prefixOrName string) (*Container, error) {
 		return containerByName, nil
 	}
 
-	containerID, indexError := daemon.idIndex.Get(prefixOrName)
-	if indexError != nil {
-		// When truncindex defines an error type, use that instead
-		if strings.Contains(indexError.Error(), "no such id") {
+	containerID, err := daemon.idIndex.Get(prefixOrName)
+	if err != nil {
+		if truncindex.IsNoSuchID(err) {
```
Comment (duglin) [inline daemon/daemon.go]: Is there a reason you created a func for this test instead of just doing a typecast check?

**Comment by duglin** [comment]
LGTM just one minor question

File: daemon/daemon.go (hunk 1 above)
Comment (runcom) [inline daemon/daemon.go]: more go'ish for a lib, like `IsNotExist` in the stdlib os pkg

**Comment by cpuguy83** [comment]
This is not quite right. There shouldn't necessarily be an error type for each error.

Take a look at https://golang.org/src/os/file_unix.go?s=3611:3650#L128
In the stdlib packages of kinds of errors, and these tend to wrap the real error and provide some context, formatting in the `Error()` function, and in some cases provide functions to see `is it this type of error?`, as in `os.IsNotExist()` (implemented here https://golang.org/src/os/error_unix.go?s=465:496#L13)

**Comment by runcom** [comment]
@cpuguy83 I took ideas exactly from there. Are you saying to compact all those 3 errors in a generic type, say TruncErr, and distinguish later in Error() output? Btw, I see those 3 errors as different kinds of errors, each with its own type. `PathError` is clearly groupable, NoSuchID and InsertionFailed aren't. I agree IsNoSuchID is redunant if it only tries one error though..

File: pkg/plugins/client.go [hunk 2]
```diff
@@ -99,9 +99,9 @@ func (c *Client) callWithRetry(serviceMethod string, args interface{}, ret inter
 		if resp.StatusCode != http.StatusOK {
 			remoteErr, err := ioutil.ReadAll(resp.Body)
 			if err != nil {
-				return &remoteError{err.Error(), serviceMethod}
+				return remoteError{err.Error(), serviceMethod}
 			}
-			return &remoteError{string(remoteErr), serviceMethod}
+			return remoteError{string(remoteErr), serviceMethod}
```
Comment (cpuguy83) [inline pkg/plugins/client.go]: How come you are casting here instead of using `err.Error()`?

Comment (runcom) [inline pkg/plugins/client.go]: I don't know, I didn't write this code, I'll change it tho

**Comment by cpuguy83** [comment]
`ErrIDAlreadyExists` and `ErrNoSuchID` are the same kind of error.
`ErrInsertionFailed` is indeed a different kind as it probably means some internal state is bad.

Also check out the net package which has some really good error handling examples (especially with formatting for the error string).

File: pkg/plugins/client.go (hunk 2 above)
Comment (runcom) [inline pkg/plugins/client.go]: no wait remoteErr is a []byte from ReadAll

Comment (cpuguy83) [inline pkg/plugins/client.go]: Oh yes, I totally cross-eyed this and saw the err.Error() in red on the left, and the green on the right thinking it was changed.

Comment (LK4D4) [inline pkg/plugins/client.go]: So weird name for []byte :)

File: pkg/truncindex/truncindex.go [hunk 3]
```diff
@@ -24,6 +24,46 @@ var (
...
+type ErrInsertionFailed struct {
+	id string
+}
+
+func (e ErrInsertionFailed) Error() string {
+	return fmt.Sprintf("failed to insert id: %s", e.id)
+}
+
+// ErrNoSuchID is returned if the TruncIndex doesn't contain the given id
+type ErrNoSuchID struct {
```
Comment (LK4D4) [inline pkg/truncindex/truncindex.go]: I can't say that I like this. It's sorta different in stdlib, you're passed that ID, why do you need it in error(remeber that truncindex is "library" and you shouldn't rely on your knowledge of calling code now)? So, I think `var ErrNoSuchID = errors.New` would be enough.

Comment (runcom) [inline pkg/truncindex/truncindex.go]: agree

File: pkg/plugins/client.go (hunk 2 above)
Comment (runcom) [inline pkg/plugins/client.go]: I'll change it

**Comment by runcom** [comment]
I'll update and reopen once done, there are  weird test errors I need to solve

**Pull request closed without merging**